src/crypto/poly1305.cpp

```diff
@@ -0,0 +1,198 @@
+  if (inlen < 16)
+    goto poly1305_donna_atmost15bytes;
+poly1305_donna_16bytes:
+  m += 16;
+  inlen -= 16;
+
+  t0 = U8TO32_LE(m - 16);
+  t1 = U8TO32_LE(m - 12);
+  t2 = U8TO32_LE(m - 8);
+  t3 = U8TO32_LE(m - 4);
+
+  h0 += t0 & 0x3ffffff;
+  h1 += ((((uint64_t)t1 << 32) | t0) >> 26) & 0x3ffffff;
+  h2 += ((((uint64_t)t2 << 32) | t1) >> 20) & 0x3ffffff;
+  h3 += ((((uint64_t)t3 << 32) | t2) >> 14) & 0x3ffffff;
+  h4 += (t3 >> 8) | (1 << 24);
+
+poly1305_donna_mul:
+  t[0] = mul32x32_64(h0, r0) + mul32x32_64(h1, s4) + mul32x32_64(h2, s3) +
+         mul32x32_64(h3, s2) + mul32x32_64(h4, s1);
+  t[1] = mul32x32_64(h0, r1) + mul32x32_64(h1, r0) + mul32x32_64(h2, s4) +
+         mul32x32_64(h3, s3) + mul32x32_64(h4, s2);
+  t[2] = mul32x32_64(h0, r2) + mul32x32_64(h1, r1) + mul32x32_64(h2, r0) +
+         mul32x32_64(h3, s4) + mul32x32_64(h4, s3);
+  t[3] = mul32x32_64(h0, r3) + mul32x32_64(h1, r2) + mul32x32_64(h2, r1) +
+         mul32x32_64(h3, r0) + mul32x32_64(h4, s4);
+  t[4] = mul32x32_64(h0, r4) + mul32x32_64(h1, r3) + mul32x32_64(h2, r2) +
+         mul32x32_64(h3, r1) + mul32x32_64(h4, r0);
+
+  h0 = (uint32_t)t[0] & 0x3ffffff;
+  c = (t[0] >> 26);
+  t[1] += c;
+  h1 = (uint32_t)t[1] & 0x3ffffff;
+  b = (uint32_t)(t[1] >> 26);
+  t[2] += b;
+  h2 = (uint32_t)t[2] & 0x3ffffff;
+  b = (uint32_t)(t[2] >> 26);
+  t[3] += b;
+  h3 = (uint32_t)t[3] & 0x3ffffff;
+  b = (uint32_t)(t[3] >> 26);
+  t[4] += b;
+  h4 = (uint32_t)t[4] & 0x3ffffff;
+  b = (uint32_t)(t[4] >> 26);
+  h0 += b * 5;
+
+  if (inlen >= 16)
+    goto poly1305_donna_16bytes;
+
+/* final bytes */
+poly1305_donna_atmost15bytes:
+  if (!inlen)
+    goto poly1305_donna_finish;
+
+  for (j = 0; j < inlen; j++)
+    mp[j] = m[j];
+  mp[j++] = 1;
+  for (; j < 16; j++)
+    mp[j] = 0;
+  inlen = 0;
+
+  t0 = U8TO32_LE(mp + 0);
+  t1 = U8TO32_LE(mp + 4);
+  t2 = U8TO32_LE(mp + 8);
+  t3 = U8TO32_LE(mp + 12);
+
+  h0 += t0 & 0x3ffffff;
+  h1 += ((((uint64_t)t1 << 32) | t0) >> 26) & 0x3ffffff;
+  h2 += ((((uint64_t)t2 << 32) | t1) >> 20) & 0x3ffffff;
+  h3 += ((((uint64_t)t3 << 32) | t2) >> 14) & 0x3ffffff;
+  h4 += (t3 >> 8);
+
+  goto poly1305_donna_mul;
+
+poly1305_donna_finish:
+  b = h0 >> 26;
+  h0 = h0 & 0x3ffffff;
+  h1 += b;
+  b = h1 >> 26;
+  h1 = h1 & 0x3ffffff;
+  h2 += b;
+  b = h2 >> 26;
+  h2 = h2 & 0x3ffffff;
+  h3 += b;
+  b = h3 >> 26;
+  h3 = h3 & 0x3ffffff;
+  h4 += b;
+  b = h4 >> 26;
+  h4 = h4 & 0x3ffffff;
+  h0 += b * 5;
+  b = h0 >> 26;
+  h0 = h0 & 0x3ffffff;
+  h1 += b;
+
+  g0 = h0 + 5;
+  b = g0 >> 26;
+  g0 &= 0x3ffffff;
+  g1 = h1 + b;
+  b = g1 >> 26;
+  g1 &= 0x3ffffff;
+  g2 = h2 + b;
+  b = g2 >> 26;
+  g2 &= 0x3ffffff;
+  g3 = h3 + b;
+  b = g3 >> 26;
+  g3 &= 0x3ffffff;
+  g4 = h4 + b - (1 << 26);
+
+  b = (g4 >> 31) - 1;
+  nb = ~b;
+  h0 = (h0 & nb) | (g0 & b);
+  h1 = (h1 & nb) | (g1 & b);
+  h2 = (h2 & nb) | (g2 & b);
+  h3 = (h3 & nb) | (g3 & b);
+  h4 = (h4 & nb) | (g4 & b);
+
+  f0 = ((h0) | (h1 << 26)) + (uint64_t)U8TO32_LE(&key[16]);
```

Comment:
Just to double check: This bit shifting takes place on a 32-bit value which is then expanded to a 64-bit type. That is the intention? Applies to `h1 << 26`, `h2 << 20`, `h3 << 14` and `h4 << 8`.

Reply:
Looks like.
This has been taken from openSSH (https://github.com/openssh/openssh-portable/blob/90e51d672711c19a36573be1785caf35019ae7a8/poly1305.c).